### ISSUE TITLE
Uncaught exception due to ETIMEDOUT

### ISSUE THREAD

**Issue by charlyz** [algolia/algoliasearch-client-js#178]
Hi,

We noticed that our API would crash from time to time due to an `ETIMEDOUT` error. We nailed down the issue to `algoliasearch` which listens the `error` event only [once](https://github.com/algolia/algoliasearch-client-js/blob/master/src/server/builds/node.js#L135) although the event seems to be called more than once (not sure why).

You can reproduce the issue by making a call to the Algolia API and dropping all the packets to the port `443` with `iptables` for example:

```
sudo iptables -A OUTPUT -p tcp -m tcp --dport 443 -j DROP
```

EDIT:

Additional information:
 - Our version is `3.8.1`
 - It crashes even with a timeout
 - When using the value `1000` for the timeout, it is actually 10 seconds

**Comment by vvo** [comment]
Hi thanks for reporting this issue, I am looking into it right now.

**Comment by vvo** [comment]
@charlyz I was not able to reproduce the issue, I think I need some more details:
- what type of calls are you doing that then leads to a double error event? Search or indexing? If you have a simple test script failing then you can put it here (without any sensible data)
- what is your nodejs version?

> It crashes even with a timeout

Do you mean that you are removing/adding a timeout option by default? If so, how?

There's a default timeout already in the library, that's why I am asking.

> When using the value 1000 for the timeout, it is actually 10 seconds

Yes the timeout is incremental given the number of hosts we try, we always try 4 hosts:
- request on first machine (tries = 0)
- timeout after 1000ms (timeout * tries + 1)
- request on second machine (tries = 1)
- timeout after 2000ms (timeout * tries + 1)
- request on third machine (tries = 2)
- timeout after 3000ms (timeout * tries + 1)
- request on third machine (tries = 3)
- timeout after 4000ms (timeout * tries + 1)

=== global timeout was 10s (1000 + 2000 + 3000 + 4000)

Default nodejs timeout is an inactivity timeout of 15s, so that if no byte is transferred for 15s then we timeout and try 15s * 2 etc..

**Comment by vvo** [comment]
Ok I think I managed to make it fail, now investigating, feel free to still answer my questions, would help

**Comment by vvo** [comment]
@charlyz Can you try this:

```sh
npm install algoliasearch@beta
```

And see if you reproduce the issue. Thanks

**Comment by charlyz** [comment]
Hi @vvo , thanks for the quick response!

I tried with `algoliasearch@beta` and I still have the same issue. Here are the [debug logs](https://gist.github.com/charlyz/390b3c3e9f523dc98ca1). If it can help, I don't have the issue by changing [this line](https://github.com/algolia/algoliasearch-client-js/blob/master/src/server/builds/node.js#L135) by `on('error', fn)`.  I then see the following multiple times then it stops:

```
algoliasearch:/1/indexes/events/query error: {"code":"ETIMEDOUT","errno":"ETIMEDOUT","syscall":"connect"}  - https://XXX-dsn.algolia.net/1/indexes/events/query +30s
```

Out of the box, I would say that when the library retries a request, it doesn't correctly close the previous socket which timeouts at some point and the error is not caught. 

To answer your questions:
 - Node version: v0.10.29
 - We're doing a search call as you can see in the logs

It also takes ~3 minutes before the OS triggers the timeout but it happens every time when dropping the packets.

Let me know if you need more information.

**Comment by vvo** [comment]
Ok thanks for the details and debugging, indeed I will switch to .on() and move on. I did not want to do it to avoid memory leaks but since we are destroying the socket and request we should be good.

What about I update the beta version, you try again, even in production and tell me if you are seeing any memory leak?

**Comment by vvo** [comment]
I was still not able to reproduce the issue, I am sure it depends on another network configuration.

Socket hangup means the other side of the request canceled it, wonder why it happens in your setup

**Comment by vvo** [comment]
As a last resort could you also do a DEBUG session with:

```
NODE_DEBUG=http DEBUG=* yourscript?
```

Do you think you could provide me a reduced test case?

**Comment by vvo** [comment]
Ok you can try again:

```sh
npm install algoliasearch@3.10.1-beta1
```

**Comment by charlyz** [comment]
I'll try the new beta. [Here](https://gist.github.com/charlyz/b7c34fa453198b72bcdb) are the logs with HTTP enabled. The use case is very basic, just a search with the new `iptables` rule. Are you on linux? Not sure if Mac has it. Also, could it be related to the `keepaliveagent`? As you said, you're abording the http request which destroys the socket. So it should be good.

**Comment by vvo** [comment]
Yes I am on linux, thanks for the new logs.

Try the new beta and let me know, then I will release it.

**Comment by vvo** [comment]
> Also, could it be related to the keepaliveagent

Maybe, hard to say.

Also maybe due to the fact that node 0.10.29 is old and this particular case you are facing was then fixed (I know it's hard to move sometimes).

But let's move on, we already spent good time and I think we got a good fix. Thanks!

**Comment by charlyz** [comment]
It seems that you're right about the Node.js version being the culprit. This guy described the [issue](https://github.com/nodejs/node-v0.x-archive/issues/6928) that we have i.e. `ECONNRESET` is triggered followed by `EDTIMEDOUT`. It seems that in the versions `0.10.xx`, `req.abort()` doesn't propagate the information down the to socket when using `https`. By adding

```
if(req.socket.socket) req.socket.socket.destroy(); 
```

before the two `req.abort()` in `node.js`, it doesn't crash. Would it make sense to use that fix for older Node versions instead of keeping sockets open?

**Comment by vvo** [comment]
> before the two req.abort() in node.js, it doesn't crash. Would it make sense to use that fix for older Node versions instead of keeping sockets open?

I am not sure since we are using keepalive. I have too little knowledge in the full net code of nodejs 0.10 to guarantee this change will not introduce something bad.

I think listening to the error event more than once still makes sense since it's not stated in nodejs docs that there should be only one error event.

And then if we see something bad is hapening with this fix we can change it.

Works for you?

**Comment by charlyz** [comment]
In my [understanding](https://nodejs.org/api/http.html#http_request_abort), `req.abort()` closes the underlying TCP socket regardless if the HTTP option `keepalive=true`. The fact that it doesn't do it with `https` is definitely a bug in `0.10`. I honestly don't think that your solution will have a negative impact in our application but explicitly closing the socket when aborting makes more sense. If you don't feel confident with that idea, no problem, we'll patch and keep the library in our repo. In any case, I greatly appreciate your help. Let me know.

**Comment by vvo** [comment]
Indeed when you think about it, if that socket failed at some point then it will fail again so better close it.

I will update the code to close this socket, and will see how to do it (I can ask the 'semver' module we have already).

Do you know if this has to be done also in node 0.12 (can you do the test?)

Thanks a lot

**Comment by charlyz** [comment]
Awesome thanks. Let me test `0.12` for you.

**Comment by charlyz** [comment]
All good for `0.12.7`, I tested without the explicit deletion of the socket and without `on('error', fn)`. `ETIMEDOUT` was not triggered, just `ECONNRESET`.

**Comment by vvo** [comment]
Ok so will release the updated client tomorrow morning CET, works for you?

**Comment by charlyz** [comment]
Perfect, thanks!

**Comment by vvo** [comment]
Was published as 3.10.1, let me know

**Comment by charlyz** [comment]
Works like a charm. Thanks and Happy Friday :)
